features/steps/thens/attributes.py

```diff
@@ -60,37 +60,124 @@ def accumulate_errors(i):
         yield from errors
 
 
-@gherkin_ifc.step('The value of attribute {attribute} must be {value}')
-@gherkin_ifc.step('The value of attribute {attribute} must be {value} {display_entity:display_entity}')
-def step_impl(context, inst, attribute, value, display_entity=0):
-    # @todo the horror and inconsistency.. should we use
-    # ast here as well to differentiate between types?
-    pred = operator.eq
-    if value == 'empty':
-        value = ()
-    elif value == 'not empty':
-        value = ()
-        pred = operator.ne
-    elif 'or' in value:
-        opts = value.split(' or ')
-        value = tuple(opts)
-        pred = misc.reverse_operands(operator.contains)
-
-    if isinstance(inst, (tuple, list)):
-        inst = inst[0]
-    attribute_value = getattr(inst, attribute, 'Attribute not found')
-    if attribute_value is None:
-        attribute_value = ()
-    if inst is None:
-        # nothing was activated by the Given criteria
-        yield ValidationOutcome(inst=inst, severity=OutcomeSeverity.EXECUTED)
-    elif not pred(attribute_value, value):
-        yield ValidationOutcome(
-            inst=inst,
-            expected=None if not value else value,
-            observed=misc.recursive_unpack_value(attribute_value),
-            severity=OutcomeSeverity.ERROR
-        )
+@gherkin_ifc.step('The value of attribute {attribute} must be {value_or_comparison_op}')
+@gherkin_ifc.step('The value of attribute {attribute} must be {value_or_comparison_op} {display_entity:display_entity}')
+@gherkin_ifc.step('The value of attribute {attribute} must be {value_or_comparison_op} the expression: {expression}')
```

Comment:
Do you think that using this with constant values directly in place of the expression would be confusing? For example, “The value of attribute <RadiusOfCurvature> must be equal to the expression: 0”?

I ask because, in ALB011, the StartRadiusOfCurvature and EndRadiusOfCurvature of IfcAlignmentHorizontalSegment need to be zero or non-zero according to its PredefinedType. This cannot be verified with the first expression because "The value of attribute <RadiusOfCurvature> must be 0.0" would compare the 0.0 float value of the IFC file with the 0.0 string value declared on the rule. It also does not consider the precision.

Reply:
> “The value of attribute must be equal to the expression: 0”

While maybe not ideal, this pattern would be fine IMHO.

Reply:
I would first try to fit it into

https://github.com/buildingSMART/ifc-gherkin-rules/blob/development/features/steps/thens/attributes.py#L63

Could be as simple as:

```
>>> import ast
>>> required = '1.0'
>>> try: required = ast.literal_eval(required)
... except: pass
...
>>> type(required)
<class 'float'>
```

But maybe also take into account tolerances when you detect you're comparing floats

Reply:
> > “The value of attribute must be equal to the expression: 0”
> 
> While maybe not ideal, this pattern would be fine IMHO.

This would be particularly useful to evaluate if StartRadiusOfCurvature and EndRadiusOfCurvature of IfcAlignmentHorizontalSegment is not zero for nonlinear PredefinedTypes, as the "not equal to" operator can be used.


> I would first try to fit it into
> 
> https://github.com/buildingSMART/ifc-gherkin-rules/blob/development/features/steps/thens/attributes.py#L63
> 
> Could be as simple as:
> 
> ```
> >>> import ast
> >>> required = '1.0'
> >>> try: required = ast.literal_eval(required)
> ... except: pass
> ...
> >>> type(required)
> <class 'float'>
> ```
> 
> But maybe also take into account tolerances when you detect you're comparing floats

The comparison using tolerances, after evaluating the string, can be done using https://github.com/buildingSMART/ifc-gherkin-rules/blob/42959f2d1b1d4defcd6fa8ec592eddb4a4f63977/features/steps/utils/geometry.py#L227

However, the expression as is would only compare using the "equal to" operator, which is not applicable to this case.